README.md
[![GitHub Actions][actions badge]][actions]
![GitHub License][license badge]

# rails_cook

A library to verify and decrypt [Ruby on Rails][] session cookies written in
[Erlang][].

[Ruby on Rails]: https://rubyonrails.org/
[Erlang]: https://www.erlang.org

## Requirements

Requires Erlang/OTP >= 22.

Erlang/OTP >= 22 is required due to the use of the new crypto API.

## Summary

Want to reuse a session created by a Ruby on Rails application for an Erlang
service, such as a web API or websocket server?

```erlang
try rails_cook:verify_and_decrypt(auto, <<"secret key base">>, <<"cookie">>) of
  {RailsVersion, CookiePayload} -> jiffy:decode(CookiePayload) % or your preferred deserialization
catch
  error:bad_sig -> "Invalid session cookie";
  error:unknown_rails_version -> "Unable to determine Rails version"
end
```

The `auto` argument will ask the library to attempt to automatically infer the
Rails version that created the cookie. Rails went though a couple of different
cookie storage formats over the years (supported are 4 and 5.1. Version 3 or
less is not supported due to the cookie payload being serialized with Ruby
object marshaling).

Instead of `auto`, you can specify the version explicitly:

```erlang
rails_cook:verify_and_decrypt(rails_5_1, <<"secret key base">>, <<"cookie">>)
```

The utilized decryption/verification and HMAC digest algorithms are
configurable, as they are in Rails, so if you changed the defaults in your Rails
setup, you should be able to get this library to decrypt your cookie. Reasonable
defaults are assumed:

| Purpose                 | Rails 4     | Rails 5     |
| ----------------------- |:-----------:|:-----------:|
| HMAC digest             | SHA-1       | SHA-1       |
| Decryption/Verification | AES-256-CBC | AES-256-GCM |

To override any property, just pass an additional proplist to any of the
functions. All exported functions have a variant that expects a proplist as
the final argument:

```erlang
try rails_cook:verify_and_decrypt(rails_4,
                                  <<"secret key base">>,
                                  <<"cookie">>,
                                  [{'hmac_hash_alg', md5}]) of
  % ...
catch
  % ...
end
```

A full list of settings is given here:

| Setting                               | Description                                                                            |
| ------------------------------------- | -------------------------------------------------------------------------------------- |
| `hmac_hash_alg`                       | Sets the HMAC digest algorithm to be used for signed cookies<br>Rails config setting: `config.action_dispatch.signed_cookie_digest`<br>Default: `sha`|
| `cipher_alg`                          | Sets the cipher to be used for decryption/verification of encrypted cookies<br>Rails config setting: `config.action_dispatch.encrypted_cookie_cipher`<br>Default: `aes_256_cbc`, `aes_256_gcm`|
| `encrypted_cookie_salt`               | Cookie salt (encrypted cookies)<br>Rails config setting: `config.action_dispatch.encrypted_cookie_salt`<br>Default: `<<"encrypted cookie">>`|
| `encrypted_signed_cookie_salt`        | Cookie salt (encrypted & signed cookies)<br>Rails config setting: `config.action_dispatch.encrypted_signed_cookie_salt`<br>Default: `<<"signed encrypted cookie">>`|
| `authenticated_encrypted_cookie_salt` | Cookie salt (authenticated & encrypted cookies)<br>Rails config setting: `config.action_dispatch.authenticated_encrypted_cookie_salt`<br>Default: `<<"authenticated encrypted cookie">>`|

## To Do

This library currently makes no effort to actually deserialize the cookie
payload. Whether it should do so, or leave that to you is being considered.

The API is not yet stable, and may change with updates.

<!-- Badges -->
[actions badge]: https://img.shields.io/github/workflow/status/vipera/rails_cook/CI?style=flat-square
[actions]: https://github.com/vipera/rails_cook/actions?query=workflow%3ACI
[license badge]: https://img.shields.io/github/license/vipera/rails_cook?style=flat-square


CHANGELOG.md
## 0.1.1 (upcoming)

### Bug fixes

- Base the Rails 4 HMAC verification's key generation on the block length of the
  chosen digest algorithm.

### Misc

- Add tests for Rails 4 encryption and verification with ciphers/digest
  algorithms other than AES-256-CBC/SHA1.

## 0.1.0 (2020-12-07)

### Feature changes

- Renamed `rails5` format to `rails_4` and added `rails_5_1` format for decoding
  authenticated & encrypted session cookies used by Rails 5.1+.
- Changed the result of any decrypt method to include the cookie format detected
  to allow disambiguation between randomly decrypted sessions for further
  processing.
- Minimal platform requirement moving up to OTP 22 due to use of new crypto
  API.

## New features

- Added support for Rails 5.1+ session decoding.
- Added `auto` format for automatically infering the Rails version from the
  cookie format.
- Added `infer_rails_version/1` for programatically inferring Rails version from
  a cookie payload.
- Added crypto properties for specifying custom salts, HMAC algorithms and
  ciphers, to support non-default Rails configurations.

## Misc

- Added config for use with rebar3.
- Added .gitignore, GitHub workflow and simple test for ensuring correct
  behaviour.
- Added types and specs for API methods.

## 0.0.1 (2019-10-10)

## New features

- Initial proof of concept library with working Rails 4 decryption and
  verification.


test/rails_cook_tests.erl
-module(rails_cook_tests).

-include_lib("eunit/include/eunit.hrl").
-include("rails_cook.hrl").

rails_cook_test_() ->
  {foreach, fun setup/0, fun teardown/1,
    [{with, [T]} || T <- [
      fun test_rails4_infer/1,
      fun test_rails51_infer/1,
      fun test_unknown_infer/1,
      fun test_verify_unknown/1,
      fun test_verify_rails4/1,
      fun test_verify_rails51/1,
      fun test_decrypt_unknown/1,
      fun test_decrypt_rails4/1,
      fun test_decrypt_rails5/1,
      fun test_verify_and_decrypt_rails4/1,
      fun test_verify_and_decrypt_rails51/1,
      fun test_verify_and_decrypt_rails51_salt_changed/1,
      fun test_verify_and_decrypt_rails51_cipher_aes192gcm/1]]}.

setup() ->
    % Uncomment to run tests with dbg:
    % dbg:tracer(),
    % dbg:p(all, call),
    % dbg:tpl(rails_cook, []),
    rails_cook.

teardown(_Module) -> ok.

-define(RAILS4_SECRET_BASE,
  <<"3431bc53c9a48f5ad58a0ba268bfc3b27e50847fc09a57128c81618b91a0de310a2fb725c8"
  "24deab3ba802642a1465235cfce9f194596fec15c2cad4ddae7353">>
).
-define(RAILS4_COOKIE_DEFAULT,
  <<"cSsvQWJKcHo2UUJ3TjZObmNyWlNxK0IyRU9OeDhTNVk0RWFrZmJ4SEFuM0w3cTI1RnlYbEZXeX"
  "NVMk9tMXorMVcvbjd6a3oydEdwZktmV2t4K1NXb2xTcWZQUEdNR1hBalM0QitBZmRJYzZkSlE3UD"
  "QxSERpeFhudGdMbXJ0WWxrSTIyYkxGNUhodmNvcWZnOFRJcWkvL1NPOFdSVThacytMZ1JlMElzVU"
  "F2TUF5VWZoYXVRZ3QyYk9udWxzTXkxUzF6cUF1NEhoRytLSlJXajc1bi9UdDF4OEgzbk1rTkR0b2"
  "tlaFFnUjlhTT0tLUVDZElxY0tTcWdHN0hMUXd1UlZ0N2c9PQ%3D%3D--5d0db878511fa0375932"
  "e23a4a19b6d6dcff5323">>
).
-define(RAILS4_COOKIE_ENCRYPTED_PAYLOAD,
  <<"q+/AbJpz6QBwN6NncrZSq+B2EONx8S5Y4EakfbxHAn3L7q25FyXlFWysU2Om1z+1W/n7zkz2tG"
  "pfKfWkx+SWolSqfPPGMGXAjS4B+AfdIc6dJQ7P41HDixXntgLmrtYlkI22bLF5Hhvcoqfg8TIqi/"
  "/SO8WRU8Zs+LgRe0IsUAvMAyUfhauQgt2bOnulsMy1S1zqAu4HhG+KJRWj75n/Tt1x8H3nMkNDto"
  "kehQgR9aM=--ECdIqcKSqgG7HLQwuRVt7g==">>
).
-define(RAILS4_COOKIE_DEFAULT_PLAINTEXT,
  <<"{\"session_id\":\"54997d3aca182c04b4c3dccb0f5d245a\",\"warden.user.user.ke"
  "y\":[[17],\"$2a$04$pX8h2AAPucIQ7ClnIiAdLe\"],\"_csrf_token\":\"j44q9drLtA4vq"
  "5z8+jhF35zBsTypbGWHVagGTRRMLdE=\"}">>
).

-define(RAILS5_SECRET_BASE,
  <<"e7e4eb74a5f53c851a3fa26561b882ab4b762567ebd67804f4feaf1bef1cd5bef2b00bf811"
  "10dcf36e965215472fe4bdf8b49149b6b6471f4e0f6f2cd2b58602">>
).
-define(RAILS5_COOKIE_DEFAULT,
  <<"Ym9OcjbK0D6fr%2B9GqKvzSknLPv8yPE%2Fa%2Fj9%2FYquApL8IuUlWeuKqYbB4P6i5mguQ93"
  "GxPtXx%2BsGasEkZlywdL6BPo9tW9UapoD5Ok%2F%2BKbm46gM9HRg%2FM7bj0YIyhSNJznU%2Fy"
  "7SXvWlcRtTEToOSPFA9e%2BTgVzFzHvJ744YW%2FOlnU5%2BrDds5vK4xFNcG6FfGUCOwoGrbDDp"
  "KjOpb3EJCoMtX%2FqKOnTpSfrE0n9bRyqQOOf694iodOF56l6eAzZP%2FZtVzBNsJRxD52igCiv9"
  "SYL1CfINm%2BXqMBJn5W%2FmBJKj0mO%2BDFpO711xmDGnItbgaY5X80QUL8qQK8Kbsfl80pyijS"
  "CvJuyXp9Tou%2BWAkSR9VZPhhexfpx2G0cOQ27%2B0ttdlCuOK8E3BcUeMQlgA8pWSqoT6m1ENUH"
  "Og%3D%3D--CkhHK5nHNHVWnWw4--XE%2BFWYSbox8iQI2gbEWnrA%3D%3D">>
).
-define(RAILS5_COOKIE_DEFAULT_PLAINTEXT,
  <<"{\"_rails\":{\"message\":\"eyJzZXNzaW9uX2lkIjoiMjBkZGYyY2U0ZGYyZDI2MTkyMDh"
  "hNDQ3ZDQ4MWFmNDUiLCJ3YXJkZW4udXNlci51c2VyLmtleSI6W1sxXSwiJDJhJDEyJDliaURzY2x"
  "hbjd5Wk43SFdxM2FJWGUiXSwiZmxhc2giOnsiZGlzY2FyZCI6W10sImZsYXNoZXMiOnsibm90aWN"
  "lIjoiU2lnbmVkIGluIHN1Y2Nlc3NmdWxseS4ifX19\",\"exp\":null,\"pur\":\"cookie._t"
  "esty_railsy_session\"}}">>
).
-define(RAILS5_COOKIE_SALT,
  <<"g2fbs3FHPPQhFlqgiM3eK2g2XpKZSWqNnx7h2IxT6%2Fg0NjS7RQPMgyHoIoP%2B1nF%2FlxlN"
  "xsQVuSF912HQtkFEM2jCld7ltjkg8k%2FxPAsNzcNfWAJ9q5%2FnpQO%2BrldusWlvp43G%2FVPZ"
  "8WyUDE%2FVdJIQvGfa%2BJf%2Fk7HIVg%2B9%2FC3%2BDtbSksEQ%2FmrND3y6VBW0jvjKfEzf9x"
  "1mzSFFcV4NQSiJdmpi6J2yAUXek22mv7x3PIEDv%2BYouI%2Fws1JbMmKIXDXWVuvrjYNwSby6Mf"
  "v3VJFq12ms%2FC02v61KTglCJs6dvr%2BFYLau1LgErF9qJ8Q%2FR2aBij6EPfZAHQaQZ9fOMusw"
  "1Xy%2BSPXYa%2F6BGk2gFRz7ra3Gx3jU8LYczUaddjytmaIYHWHGW%2FMPj0VkYxeE06JJCu%2Bf"
  "62eWQPUyXg%3D%3D--x9mMJjLrWdekMoXx--XqaS3xAztPWibBbW30NMnw%3D%3D">>
).
-define(RAILS5_COOKIE_SALT_PLAINTEXT,
  <<"{\"_rails\":{\"message\":\"eyJzZXNzaW9uX2lkIjoiYTRjMjIyYTQ0MGJlZGMzNjlkNzN"
  "kODNhOWZiMGQzZDIiLCJ3YXJkZW4udXNlci51c2VyLmtleSI6W1sxXSwiJDJhJDEyJDliaURzY2x"
  "hbjd5Wk43SFdxM2FJWGUiXSwiZmxhc2giOnsiZGlzY2FyZCI6W10sImZsYXNoZXMiOnsibm90aWN"
  "lIjoiU2lnbmVkIGluIHN1Y2Nlc3NmdWxseS4ifX19\",\"exp\":null,\"pur\":\"cookie._t"
  "esty_railsy_session\"}}">>
).
-define(RAILS5_COOKIE_AES192GCM,
  <<"PcDbAf9Nq1mMn%2Bw8YeD8rXGDi9mQTVegAHruS2QJsgiqd%2BqF5YyB3CmlQg%2B6sN36lf2d"
  "3fWX%2Fjd%2FbXvlfBYaBExy1Yua%2BQh9Sc89lKtZ5flxx%2BvQ6ylHCDK1IWT%2BeE71kCe6er"
  "jmmrJ3ZvO1ckU1LaZ2spCIUrGa%2BNUrgVp3iT0IH5aue6uwsCahcJmRBgHWLe82jTVy%2BLazlp"
  "VzNc0%2FxY98wnjisjiPDgmoebJkb%2FRX1UGoA1DHwsMgV1yPn6EX8HgW%2BMN6GE%2F6llXDAn"
  "nPoM8%2BS4%2BpPsi6%2FMp4uZGEQvKH4gv59RlO5qdmkl9jN7RQl%2B2Q0BJtCOz5U1z8rxM%2F"
  "Wte0I%2BOGKTq13bkMHCmxziyYt5QHRqGcvm9eVBMsOvyoXjNBME7%2BJerE17tRsD4BSCbwV7tz"
  "TkosxQ%3D%3D--t%2BHbujcf4Xx64I9B--Lfqo%2FsZqBTApqgQMjvd9AQ%3D%3D">>
).
-define(RAILS5_COOKIE_AES192GCM_PLAINTEXT,
  <<"{\"_rails\":{\"message\":\"eyJzZXNzaW9uX2lkIjoiNDU2N2JmYjY5YzFkNDQ2YWRjODk"
  "wZmNjZjU5Y2VlMzIiLCJ3YXJkZW4udXNlci51c2VyLmtleSI6W1sxXSwiJDJhJDEyJDliaURzY2x"
  "hbjd5Wk43SFdxM2FJWGUiXSwiZmxhc2giOnsiZGlzY2FyZCI6W10sImZsYXNoZXMiOnsibm90aWN"
  "lIjoiU2lnbmVkIGluIHN1Y2Nlc3NmdWxseS4ifX19\",\"exp\":null,\"pur\":\"cookie._t"
  "esty_railsy_session\"}}">>
).

test_rails4_infer(Mod) ->
  Version = Mod:infer_rails_version(?RAILS4_COOKIE_DEFAULT),
  ?assertEqual(rails_4, Version, "Correctly detects cookie format as Rails 4").

test_rails51_infer(Mod) ->
  Version = Mod:infer_rails_version(?RAILS5_COOKIE_DEFAULT),
  ?assertEqual(rails_5_1, Version,
    "Correctly detects cookie format as Rails 5.1").

test_unknown_infer(Mod) ->
  Version = Mod:infer_rails_version(<<"Not known">>),
  ?assertEqual(unknown, Version, "Infers unknown if not a recognized format").

test_verify_unknown(Mod) ->
  ?assertError(unknown_rails_version,
    Mod:verify(unknown, <<"secret key base">>, <<"cookie">>),
    "Cannot verify a cookie for an unknown Rails version").

test_verify_rails4(Mod) ->
  ?assertEqual({ok, ?RAILS4_COOKIE_ENCRYPTED_PAYLOAD},
    Mod:verify(rails_4, ?RAILS4_SECRET_BASE, ?RAILS4_COOKIE_DEFAULT),
    "Correctly verifies a Rails 4 session cookie").

test_verify_rails51(Mod) ->
  Cookie = <<"can be whatever">>,
  ?assertEqual({ok, Cookie},
    Mod:verify(rails_5_1, <<"secret key base">>, Cookie),
    "Rails 5.1 uses a null verifier").

test_decrypt_unknown(Mod) ->
  ?assertError(unknown_rails_version,
    Mod:decrypt(unknown, <<"secret key base">>, <<"cookie">>),
    "Cannot decrypt a cookie for an unknown Rails version").

test_decrypt_rails4(Mod) ->
  {CookieFormat, Result} = Mod:decrypt(rails_4, ?RAILS4_SECRET_BASE,
    ?RAILS4_COOKIE_ENCRYPTED_PAYLOAD),
  ?assertEqual(rails_4, CookieFormat,
    "Correctly detects cookie format as Rails 4"),
  ?assertEqual(?RAILS4_COOKIE_DEFAULT_PLAINTEXT, Result,
    "Correctly decrypts a Rails 4 session cookie after validation").

test_decrypt_rails5(Mod) ->
  {CookieFormat, Result} = Mod:decrypt(rails_5_1, ?RAILS5_SECRET_BASE,
    ?RAILS5_COOKIE_DEFAULT),
  ?assertEqual(rails_5_1, CookieFormat,
    "Correctly detects cookie format as Rails 5.1"),
  ?assertEqual(?RAILS5_COOKIE_DEFAULT_PLAINTEXT, Result,
    "Correctly decrypts a Rails 5.1 session cookie").

test_verify_and_decrypt_rails4(Mod) ->
  {CookieFormat, Result} = Mod:verify_and_decrypt(
    auto, ?RAILS4_SECRET_BASE, ?RAILS4_COOKIE_DEFAULT),
  ?assertEqual(rails_4, CookieFormat,
    "Correctly detects cookie format as Rails 4"),
  ?assertEqual(?RAILS4_COOKIE_DEFAULT_PLAINTEXT, Result,
    "Correctly verifies and decrypts a Rails 4 session with the valid base").

test_verify_and_decrypt_rails51(Mod) ->
  {CookieFormat, Result} = Mod:verify_and_decrypt(
    auto, ?RAILS5_SECRET_BASE, ?RAILS5_COOKIE_DEFAULT),
  ?assertEqual(rails_5_1, CookieFormat,
    "Correctly detects cookie format as Rails 5.1"),
  ?assertEqual(?RAILS5_COOKIE_DEFAULT_PLAINTEXT, Result,
    "Correctly verifies and decrypts a Rails 5.1 session with the valid base"),

  ?assertError(bad_sig,
    Mod:verify_and_decrypt(auto, <<"bad secret base">>, ?RAILS5_COOKIE_DEFAULT),
    "Errors if using incorrect secret key base").

test_verify_and_decrypt_rails51_salt_changed(Mod) ->
  {CookieFormat, Result} = Mod:verify_and_decrypt(
    auto, ?RAILS5_SECRET_BASE, ?RAILS5_COOKIE_SALT,
    [{authenticated_encrypted_cookie_salt, <<"other salt">>}]),
  ?assertEqual(rails_5_1, CookieFormat,
    "Correctly detects cookie format as Rails 5.1"),
  ?assertEqual(?RAILS5_COOKIE_SALT_PLAINTEXT, Result,
    "Correctly verifies and decrypts a Rails 5.1 session with custom salt").

test_verify_and_decrypt_rails51_cipher_aes192gcm(Mod) ->
  {CookieFormat, Result} = Mod:verify_and_decrypt(
    auto, ?RAILS5_SECRET_BASE, ?RAILS5_COOKIE_AES192GCM,
    [{cipher_alg, aes_192_gcm}]),
  ?assertEqual(rails_5_1, CookieFormat,
    "Correctly detects cookie format as Rails 5.1"),
  ?assertEqual(?RAILS5_COOKIE_AES192GCM_PLAINTEXT, Result,
    "Correctly verifies and decrypts a Rails 5.1 session with a different cipher").


src/rails_cook.erl
%%% rails_cook
%%% Rails cookie utilities

-module(rails_cook).
-export([infer_rails_version/1,
         infer_rails_version/2,
         verify/3,
         verify/4,
         decrypt/3,
         decrypt/4,
         verify_and_decrypt/3,
         verify_and_decrypt/4]).

-define(ENCRYPTED_COOKIE_SALT, <<"encrypted cookie">>).
-define(ENCRYPTED_SIGNED_COOKIE_SALT, <<"signed encrypted cookie">>).
-define(AUTHENTICATED_ENCRYPTED_COOKIE_SALT, <<"authenticated encrypted cookie">>).

-type rails_version() :: rails_4 | rails_5_1.
-type cookie() :: binary().
-type secret_key_base() :: binary().

-type crypto_opts() :: [ crypto_opt() ].
-type crypto_opt() :: { hmac_hash_alg, atom() }
                    | { cipher_alg, atom() }
                    | { encrypted_cookie_salt, binary() }
                    | { encrypted_signed_cookie_salt, binary() }
                    | { authenticated_encrypted_cookie_salt, binary() }.

-type verify_result() :: verify_error() | { ok, binary() }.
-type verify_error() :: bad_sig.

-type decrypt_result() :: { rails_version(), binary() }.

%% Compatibility macros for supporting multiple OTP versions
-if(?OTP_RELEASE >= 23).
  %% OTP 23 provides crypto_one_time/5 with 'pcks_padding' padding type.
  -define(rails4Decrypt(CipherAlg, EncCookieKey, IV, Ciphertext),
    begin
    ((fun () ->
        case crypto:crypto_one_time(CipherAlg, EncCookieKey, IV, Ciphertext,
                                    [{encrypt, false}, {padding, pkcs_padding}]) of
          {_, _, Description} -> {error, Description};
          Result -> Result
        end
      end)())
    end).
-elif(?OTP_RELEASE >= 22).
  %% Workaround for OTP 22 - manually remove PCKS7 padding after decryption
  -define(rails4Decrypt(CipherAlg, EncCookieKey, IV, Ciphertext),
    begin
    ((fun () ->
        case crypto:crypto_one_time(CipherAlg, EncCookieKey, IV, Ciphertext, false) of
          {_, _, Description} -> {error, Description};

          % Manually remove PKCS7 padding
          Result -> binary:part(Result, 0, byte_size(Result) - binary:last(Result))
        end
      end)())
    end).
-else.
  -error("Minimum of Erlang/OTP 22 is required to use rails_cook.").
-endif.

%%% PUBLIC API

-spec infer_rails_version(Cookie) -> rails_version()
      when Cookie :: cookie().

%% Attempts to automatically determine the rails version from the cookie format
infer_rails_version(Cookie) ->
  infer_rails_version(Cookie, []).

-spec infer_rails_version(Cookie, CryptoOpts) -> rails_version()
      when Cookie :: cookie(),
           CryptoOpts :: crypto_opts().

%% Attempts to automatically determine the rails version from the cookie format
infer_rails_version(Cookie, _Properties) when is_binary(Cookie) ->
  CookieParts = binary:split(uri_decode(Cookie), <<"--">>, [global]),
  case erlang:length(CookieParts) - 1 of
    1 -> rails_4;
    2 -> rails_5_1;
    _ -> unknown
  end.

-spec decrypt(CookieFormat, SecretKeyBase, Cookie) -> decrypt_result()
      when CookieFormat :: rails_version() | auto,
           SecretKeyBase :: secret_key_base(),
           Cookie :: cookie().

decrypt(RailsVersion, RailsKey, Cookie) ->
  decrypt(RailsVersion, RailsKey, Cookie, []).

-spec decrypt(CookieFormat, SecretKeyBase, Cookie, CryptoOpts) -> decrypt_result()
      when CookieFormat :: rails_version() | auto,
           SecretKeyBase :: secret_key_base(),
           Cookie :: cookie(),
           CryptoOpts :: crypto_opts().

%% Automatically infers Rails version from the cookie format and decrypts the
%% cookie
decrypt(auto, RailsKey, Cookie, Properties) ->
  RailsVersion = infer_rails_version(Cookie),
  decrypt(RailsVersion, RailsKey, Cookie, Properties);

decrypt(unknown, _RailsKey, _Cookie, _Properties) ->
  erlang:error(unknown_rails_version);

%% Since 5.1.0, Rails is leveraging AEAD (AES-256-GCM cipher) for faster
%% encryption/decryption and verification, in a single step.
%%
%% <cookie-payload> ::= <cipher-text> "--" <cipher-iv> "--" <cipher-auth-tag>
decrypt(rails_5_1, RailsKey, Cookie, Properties) ->
  [Ciphertext, IV, AuthTag] = lists:map(fun base64:decode/1,
                                        binary:split(uri_decode(Cookie),
                                                     <<"--">>, [global])),
  case byte_size(AuthTag) of
    16 -> ok;
    _ -> erlang:error(invalid_auth_tag_size)
  end,
  CookieSalt = case get_authenticated_encrypted_cookie_salt(Properties) of
    undefined -> ?AUTHENTICATED_ENCRYPTED_COOKIE_SALT;
    Salt -> Salt
  end,
  CipherAlg = case get_cipher_alg(Properties) of
    undefined -> aes_256_gcm;
    Alg -> Alg
  end,
  {ok, KeyLength} = maps:find(key_length, crypto:cipher_info(CipherAlg)),
  EncCookieKey = get_pbkdf2_key(RailsKey, CookieSalt, KeyLength),
  case crypto:crypto_one_time_aead(CipherAlg, EncCookieKey, IV,
                                   Ciphertext, <<>>, AuthTag, false) of
    error -> erlang:error(bad_sig);
    Plaintext -> {rails_5_1, Plaintext}
  end;

%% Since Rails 4.0, the cookie data is base64 encoded and is a PAYLOAD--IV pair,
%% where payload is encrypted with AES-256-CBC block cipher.
%%
%% <verified-payload> ::= <cipher-text> "--" <cipher-iv>
decrypt(rails_4, RailsKey, Cookie, Properties) ->
  [Ciphertext, IV] = lists:map(fun base64:decode/1,
                               binary:split(uri_decode(Cookie), <<"--">>)),
  CookieSalt = case get_encrypted_cookie_salt(Properties) of
    undefined -> ?ENCRYPTED_COOKIE_SALT;
    Salt -> Salt
  end,
  CipherAlg = case get_cipher_alg(Properties) of
    undefined -> aes_256_cbc;
    Alg -> Alg
  end,
  {ok, KeyLength} = maps:find(key_length, crypto:cipher_info(CipherAlg)),
  EncCookieKey = get_pbkdf2_key(RailsKey, CookieSalt, KeyLength),
  case ?rails4Decrypt(CipherAlg, EncCookieKey, IV, Ciphertext) of
    error -> erlang:error(cannot_decrypt);
    Plaintext -> {rails_4, Plaintext}
  end.

-spec verify(CookieFormat, SecretKeyBase, Cookie) -> verify_result()
      when CookieFormat :: rails_version() | auto,
           SecretKeyBase :: secret_key_base(),
           Cookie :: cookie().

verify(RailsVersion, RailsKey, Cookie) ->
  verify(RailsVersion, RailsKey, Cookie, []).

-spec verify(CookieFormat, SecretKeyBase, Cookie, CryptoOpts) -> verify_result()
      when CookieFormat :: rails_version() | auto,
           SecretKeyBase :: secret_key_base(),
           Cookie :: cookie(),
           CryptoOpts :: crypto_opts().

%% Automatically infers Rails version from the cookie format and verifies the
%% cookie
verify(auto, RailsKey, Cookie, Properties) ->
  RailsVersion = infer_rails_version(Cookie),
  verify(RailsVersion, RailsKey, Cookie, Properties);

verify(unknown, _RailsKey, _Cookie, _Properties) ->
  erlang:error(unknown_rails_version);

%% For Rails >= 5.1.0 A "null verifier" is used, i.e. no verification is done by
%% hand, as it is handled by the cipher. The payload is base64 encoded.
verify(rails_5_1, _RailsKey, Cookie, _Properties) ->
  {ok, Cookie};

%% Rails < 5.1.0 uses a HMAC to prove the authenticity of a cookie payload
%%
%% <cookie-payload> ::= <verified-payload-b64> "--" <hmac>
verify(rails_4, RailsKey, Cookie, Properties) ->
  %% Rails 4 cookie is HMAC-signed at the outmost layer: PAYLOAD--SIGNATURE
  [CookieData, Digest] = binary:split(uri_decode(Cookie), <<"--">>),
  %% Verify that the signature is correct
  CookieSalt = case get_encrypted_signed_cookie_salt(Properties) of
    undefined -> ?ENCRYPTED_SIGNED_COOKIE_SALT;
    Salt -> Salt
  end,
  HmacHashAlg = case get_hmac_hash_alg(Properties) of
    undefined -> sha;
    Alg -> Alg
  end,
  {ok, HashBlockSize} = maps:find(block_size, crypto:hash_info(HmacHashAlg)),
  SigEncCookieKey = get_pbkdf2_key(RailsKey, CookieSalt, HashBlockSize),
  <<Hmac:160/integer>> = crypto:mac(hmac, HmacHashAlg, SigEncCookieKey, CookieData),
  HexHmac = list_to_binary(lists:flatten(io_lib:format("~40.16.0b", [Hmac]))),
  if
    HexHmac =:= Digest -> % signature OK
      {ok, base64:decode(CookieData)};
    HexHmac =/= Digest -> % invalid signature
      bad_sig
  end.

-spec verify_and_decrypt(CookieFormat, SecretKeyBase, Cookie, CryptoOpts) -> decrypt_result()
      when CookieFormat :: rails_version() | auto,
           SecretKeyBase :: secret_key_base(),
           Cookie :: cookie(),
           CryptoOpts :: crypto_opts().

%% Automatically infers Rails version from the cookie format, then verifies and
%% decrypts the cookie
verify_and_decrypt(auto, RailsKey, Cookie, Properties) ->
  RailsVersion = infer_rails_version(Cookie),
  verify_and_decrypt(RailsVersion, RailsKey, Cookie, Properties);

verify_and_decrypt(unknown, _RailsKey, _Cookie, _Properties) ->
  erlang:error(unknown_rails_version);

verify_and_decrypt(RailsVersion, RailsKey, Cookie, Properties) ->
  case verify(RailsVersion, RailsKey, Cookie, Properties) of
    {ok, CookieData} ->
      decrypt(RailsVersion, RailsKey, CookieData, Properties);
    Reason ->
      erlang:error(Reason)
  end.

-spec verify_and_decrypt(CookieFormat, SecretKeyBase, Cookie) -> decrypt_result()
      when CookieFormat :: rails_version() | auto,
           SecretKeyBase :: secret_key_base(),
           Cookie :: cookie().

verify_and_decrypt(RailsVersion, RailsKey, Cookie) ->
  verify_and_decrypt(RailsVersion, RailsKey, Cookie, []).

%%% PRIVATE FUNCTIONS

get_encrypted_cookie_salt(Properties) ->
  proplists:get_value(encrypted_cookie_salt, Properties).

get_encrypted_signed_cookie_salt(Properties) ->
  proplists:get_value(encrypted_signed_cookie_salt, Properties).

get_authenticated_encrypted_cookie_salt(Properties) ->
  proplists:get_value(authenticated_encrypted_cookie_salt, Properties).

get_cipher_alg(Properties) ->
  proplists:get_value(cipher_alg, Properties).

get_hmac_hash_alg(Properties) ->
  proplists:get_value(hmac_hash_alg, Properties).

get_pbkdf2_key(Secret, Salt, DerivedKeyLength) ->
  Iterations = 1000,
  {ok, Key} = pbkdf2:pbkdf2(sha, Secret, Salt, Iterations, DerivedKeyLength),
  Key.

%% This implementation is based on http_uri:decode(), because there is
%% no direct alternative in the recommended uri_string module.
%% cf. http://erlang.org/pipermail/erlang-questions/2020-March/099207.html
%% @private
uri_decode(String) when is_binary(String) ->
    do_decode_binary(String).

do_decode_binary(<<$%, Hex:2/binary, Rest/bits>>) ->
    <<(binary_to_integer(Hex, 16)), (do_decode_binary(Rest))/binary>>;
do_decode_binary(<<First:1/binary, Rest/bits>>) ->
    <<First/binary, (do_decode_binary(Rest))/binary>>;
do_decode_binary(<<>>) ->
    <<>>.
